src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,189 @@
+    IdTable& resultTable,
+    sparqlExpression::EvaluationContext& evaluationContext,
+    const GroupBy& groupBy) {
+  resultTable.emplace_back();
+  for (size_t rowIndex = 0; rowIndex < groupByCols_.size(); ++rowIndex) {
+    resultTable.back()[rowIndex] = currentGroupBlock_.at(rowIndex).second;
+  }
+
+  evaluationContext._beginIndex = resultTable.size() - 1;
+  evaluationContext._endIndex = resultTable.size();
+
+  for (auto& alias : aggregateAliases_) {
+    groupBy.evaluateAlias(alias, &resultTable, evaluationContext,
+                          aggregationData_, &localVocab_);
+  }
+  resetAggregationData();
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::processNextBlock(
+    sparqlExpression::EvaluationContext& evaluationContext, size_t beginIndex,
+    size_t endIndex) {
+  size_t blockSize = endIndex - beginIndex;
+  evaluationContext._beginIndex = beginIndex;
+  evaluationContext._endIndex = endIndex;
+
+  for (auto& aggregateAlias : aggregateAliases_) {
+    for (auto& aggregate : aggregateAlias.aggregateInfo_) {
+      // Evaluate child expression on block
+      auto exprChildren = aggregate.expr_->children();
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
+
+      std::visit(
+          [this, blockSize, &evaluationContext,
+           &aggregate]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), blockSize, &evaluationContext);
+
+            for (const auto& val : generator) {
+              addToAggregateFunction(aggregate.aggregateDataIndex_, val,
+                                     &evaluationContext);
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
+}
+
+// _____________________________________________________________________________
+
+template <size_t NUM_GROUP_COLUMNS>
+void LazyGroupBy<NUM_GROUP_COLUMNS>::populateGroupBlock(const IdTable& idTable,
+                                                        size_t row) {
+  if (currentGroupBlock_.empty()) {
+    for (size_t col : groupByCols_) {
+      currentGroupBlock_.emplace_back(col, idTable(0, col));
+    }
+  } else {
+    for (auto& columnPair : currentGroupBlock_) {
+      columnPair.second = idTable(row, columnPair.first);
+    }
+  }
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_GROUP_COLUMNS>
+bool LazyGroupBy<NUM_GROUP_COLUMNS>::rowMatchesCurrentBlock(
+    const IdTable& idTable, size_t row) {
+  return std::all_of(currentGroupBlock_.begin(), currentGroupBlock_.end(),
+                     [&](const auto& columns) {
+                       return idTable(row, columns.first) == columns.second;
+                     });
```

Comment:
```suggestion
  return std::ranges::all_of(currentGroupBlock_,
                     [&](const auto& colIdxAndValue) {
                       return idTable(row, colIdxAndValue.first) == colIdxAndValue.second;
                     });
```

Reply:
But I think we shouldd (to optimize for large groups)
add a function `findIndexOfFirstMismatchingRow`, which then does the searching columnwise.